packages/gatsby/src/utils/worker/pool.ts

```diff
@@ -69,7 +73,29 @@ export async function runQueriesInWorkersQueue(
   }
 
   await Promise.all(promises)
+  await Promise.all(pool.all.saveQueries())
+  activity.end()
+}
+
+export async function mergeWorkerState(pool: GatsbyWorkerPool): Promise<void> {
+  const activity = reporter.activityTimer(`Merge worker state`)
+  activity.start()
+
+  const workerQueryState: IMergeWorkerQueryState["payload"] = []
 
+  for (const { workerId } of pool.getWorkerInfo()) {
+    const state = loadPartialStateFromDisk([`queries`], String(workerId))
+    const queryStateChunk = state.queries as IGatsbyState["queries"]
+    if (queryStateChunk) {
+      // When there are too little queries, some worker can be inactive and its state is empty
+      workerQueryState.push({ workerId, queryStateChunk })
+      await new Promise(resolve => process.nextTick(resolve))
+    }
+  }
+  store.dispatch({
+    type: `MERGE_WORKER_QUERY_STATE`,
+    payload: workerQueryState,
+  })
```

Comment:
TODO: after some thinking, we should dispatch a separate action per worker (given how large data dependencies can be)